PTM Vehicle Allocation

bccheung opened the pull request.

TianYangLin committed:
Initital commit of tour_mode_choice_reallocation

TianYangLin committed:
- Minor fixes to both overlap identification and tour mode reallocation
- Add assertions of identified tour overlaps and reallocated tour modes
- Assign reallocated logsums to final reallocated tours
- Change output columns from tour mode reallocation to align with "tours" columns

TianYangLin committed:
Initial commit of pytest module for tour mode choice reallocation

TianYangLin committed:
Clean up of auto reallocation model

TianYangLin committed:
Add auto ownership vs. tour number check to identify_auto_overallocation

TianYangLin committed:
Add check for defined mode choice logsum column name to ensure logsum is available for use in auto reallocation
Add read in of column names for pre-reallocation mode choice and logsums from module yaml file
Add check for defined mode choice logsum column from initial tour mode choice module

TianYangLin committed:
-Change check for mode choice logsum column from tour mode choice module from checking for column name in tours table to checking for column definition setting in tour mode choice module settings
-Remove redundant check for mode choice logsum column before writing old logsum column

TianYangLin committed:
- Change to time grid based method for identifiying tour overlaps
- Separate find_overlap from main identification function

TianYangLin committed:
-Separate test cases into more, smaller, more specific test cases
-Change initialization of no_auto_tours test fixture to be more efficient
-Add test case descriptions to household test fixture

TianYangLin committed:
-Moving the tour identification in front of no auto tour mode choice run

TianYangLin committed:
- Break up identify_auto_overallocation
- Cleanup, doc string completion and additional comments

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+        Household table with auto sufficiency column added
+
+    """
+    households['auto_sufficiency'] = 'suff'
```
@TianYangLin Does this column not exist in the households table by the time the model gets to tour mode choice? Other parts of the code seems to indicate the presence of a `car_sufficiency` column.

TianYangLin replied:
There are references to a "car_sufficiency" in a commented out section of joint_tour_participation and in expressions in some csvs, but is not found used in the model.

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+    """
+    households['auto_sufficiency'] = 'suff'
+    households.loc[
+        households.auto_ownership.between(0, households.num_drivers, 'neither'), 'auto_sufficiency'] = 'insuff'
```
@TianYangLin Can we double check how auto sufficiency is defined in the rest of the model? I think in other parts, it's based on number vehicles vs number of workers. What we did makes sense, just want to make note of how its defined in the rest of the model.

TianYangLin replied:
See reply to comment above on line 190

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+
+    """
+    tour_time_grid = pd.DataFrame({'household_id': filtered_tours['household_id']})
+    for i in range(1, 49):
```
@TianYangLin Are the time intervals always 30-min intervals? If the duration of each time interval is configurable in ActivitySim, then we cannot hard code the number of intervals in this module.

TianYangLin replied:
The 30-min intervals are not hard coded into the model. The time interval resolution of the model is read in implicitly through the tour departure and duration alternatives (tdd_alt). Will change this section to use the tdd_alt.

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+
+    """
+    # Find which intervals have overlaps greater than the number of autos the household owns
+    overallocation_condition = tours[[*range(1, 49)]].sum() > auto_ownership
```
@TianYangLin same comment as line 254

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+    """
+    # Find which intervals have overlaps greater than the number of autos the household owns
+    overallocation_condition = tours[[*range(1, 49)]].sum() > auto_ownership
+    overallocated_intervals = tours[[*range(1, 49)]].loc[:, overallocation_condition]
```
@TianYangLin same comment as line 254

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+    tours_insuff_drv = tours[
+        tours.household_id.isin(households_insuff.index) &
+        tours.person_id.isin(person_drivers.index) &
+        (tours.tour_mode == 'DRIVEALONEFREE')]
```
@TianYangLin We should probably also look for `DRIVEALONEPAY` for the sake of completeness, even if it's not used in PTM

TianYangLin replied:
Will add DRIVEALONEPAY

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+    tours_overallocated = filtered_tours.loc[overlap_tour_ids].copy()
+
+    # Assert that only auto-based tours are tagged
+    assert (tours_overallocated.tour_mode == 'DRIVEALONEFREE').all()
```
@TianYangLin same comment as Line 223

bccheung commented on activitysim/abm/models/tour_mode_choice_reallocation.py:
```diff
@@ -0,0 +1,653 @@
+    reallocated_tours.loc[reallocated_tours.reallocation_flag, 'mode_choice_logsum'] = reallocated_tours.no_auto_mc_logsum
+
+    # Assert reallocated tours no longer uses auto-based modes
+    assert (reallocated_tours.loc[reallocated_tours.reallocation_flag, 'tour_mode'] != 'DRIVEALONEFREE').all()
```
@TianYangLin same comment as Line 223

TianYangLin committed:
Minor update to tests to match function names

TianYangLin committed:
- Add DRIVEALONEPAY (currently not in use) to the list of modes to be filtered for in auto overallocation checks
- Add add_alts to the list of inputs to the module
- Read in the time interval resolution from add_alts for  auto overallocation identification and more specifically time grid creation
